File: bindings/ts/test/pubsub.test.ts
// Pub/sub conformance (contract §3.11 / §6): message + pattern round-trips on
// the named embedded bus AND remote; anonymous mem:// rejected; RESP3 push
// via hello3 (remote); read-timeout bounds a blocking recv. node --test + bun.

import { test, before, after } from "node:test";
import assert from "node:assert/strict";
import { connect, Subscriber, UnsupportedError } from "../src/index.ts";
import { spawnServer, uniqueMemUrl, s, type Server } from "./harness.ts";

let server: Server;
before(async () => {
  server = await spawnServer();
});
after(() => server?.close());

function pubsubUrls(): Array<[string, string]> {
  // [name, url]; embedded uses a unique named bus, remote the shared server.
  return [
    ["embedded", uniqueMemUrl()],
    ["remote", server.url],
  ];
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

test("publish → subscriber receives Message (both backends)", async () => {
  for (const [, url] of pubsubUrls()) {
    const sub = await Subscriber.connectChannels(url, "news");
    const ack = await sub.recv();
    assert.equal(ack.kind, "subscribe");
    const pub = await connect(url);
    await sleep(50);
    await pub.publish("news", "hello");
    const [ch, payload] = await sub.recvMessage();
    assert.equal(s(ch), "news");
    assert.equal(s(payload), "hello");
    sub.close();
    pub.close();
  }
});

test("pattern subscribe receives Pmessage (both backends)", async () => {
  for (const [, url] of pubsubUrls()) {
    const sub = await Subscriber.connect(url);
    await sub.psubscribe("ne*");
    const ack = await sub.recv();
    assert.equal(ack.kind, "psubscribe");
    const pub = await connect(url);
    await sleep(50);
    await pub.publish("news", "world");
    const [ch, payload] = await sub.recvMessage();
    assert.equal(s(ch), "news");
    assert.equal(s(payload), "world");
    sub.close();
    pub.close();
  }
});

test("anonymous mem:// Subscriber → Unsupported", async () => {
  await assert.rejects(Subscriber.connect("mem://"), (e) => e instanceof UnsupportedError);
});

test("remote hello3 upgrades to RESP3 push; recv handles it", async () => {
  const sub = await Subscriber.connect(server.url);
  await sub.hello3();
  await sub.subscribe("rt");
  const ack = await sub.recv();
  assert.equal(ack.kind, "subscribe");
  const pub = await connect(server.url);
  await sleep(50);
  await pub.publish("rt", "v3");
  const [ch, payload] = await sub.recvMessage();
  assert.equal(s(ch), "rt");
  assert.equal(s(payload), "v3");
  sub.close();
  pub.close();
});

test("embedded hello3 → Unsupported", async () => {
  const sub = await Subscriber.connect(uniqueMemUrl());
  await assert.rejects(sub.hello3(), (e) => e instanceof UnsupportedError);
  sub.close();
});

test("read-timeout bounds a blocking recv (both backends)", async () => {
  for (const [, url] of pubsubUrls()) {
    const sub = await Subscriber.connectChannels(url, "quiet");
    await sub.recv(); // drain ack
    sub.setReadTimeout(80);
    const start = Date.now();
    await assert.rejects(sub.recv(), (e) => (e as { kind?: string }).kind === "timedOut");
    assert.ok(Date.now() - start < 2000, "timeout bounded the recv");
    sub.close();
  }
});

File: bindings/go/pubsub_test.go
package kevy

import (
	"fmt"
	"testing"
	"time"
)

// Pub/sub round-trip (contract §3.11, §6). Named embedded bus + remote.

func pubsubBackends(t *testing.T) []backend {
	// Same rule as bothBackends: the embedded half exists only where the
	// engine was linked in.
	backends := []backend{}
	if openEmbeddedStore != nil {
		backends = append(backends, backend{
			name: "embedded",
			url: func(t *testing.T) (string, func()) {
				return fmt.Sprintf("mem://ps-%d", uniqueID()), func() {}
			},
		})
	}
	return append(backends, []backend{
		{
			name: "remote",
			url: func(t *testing.T) (string, func()) {
				s := spawnServer(t)
				return s.url(), func() {}
			},
		},
	}...)
}

func TestPubSubMessage(t *testing.T) {
	for _, bk := range pubsubBackends(t) {
		t.Run(bk.name, func(t *testing.T) {
			url, cleanup := bk.url(t)
			defer cleanup()

			sub, err := SubscriberConnectChannels(url, b("news"))
			if err != nil {
				t.Fatal(err)
			}
			defer sub.Close()
			// First recv is the subscribe ack.
			ev, err := sub.Recv()
			if err != nil || ev.Kind != EventSubscribe {
				t.Fatalf("subscribe ack: %+v %v", ev, err)
			}

			pub := mustConnect(t, url)
			// Give the remote SUBSCRIBE time to register server-side.
			time.Sleep(50 * time.Millisecond)
			if _, perr := pub.Publish(bg, b("news"), b("hello")); perr != nil {
				t.Fatal(perr)
			}

			ch, payload, rerr := sub.RecvMessage()
			if rerr != nil || string(ch) != "news" || string(payload) != "hello" {
				t.Fatalf("RecvMessage=%q %q %v", ch, payload, rerr)
			}
		})
	}
}

func TestPubSubPattern(t *testing.T) {
	for _, bk := range pubsubBackends(t) {
		t.Run(bk.name, func(t *testing.T) {
			url, cleanup := bk.url(t)
			defer cleanup()

			sub, err := SubscriberConnect(url)
			if err != nil {
				t.Fatal(err)
			}
			defer sub.Close()
			if err := sub.Psubscribe(b("ne*")); err != nil {
				t.Fatal(err)
			}
			ev, err := sub.Recv() // psubscribe ack
			if err != nil || ev.Kind != EventPsubscribe {
				t.Fatalf("psubscribe ack: %+v %v", ev, err)
			}

			pub := mustConnect(t, url)
			time.Sleep(50 * time.Millisecond)
			_, _ = pub.Publish(bg, b("news"), b("world"))

			ch, payload, rerr := sub.RecvMessage()
			if rerr != nil || string(ch) != "news" || string(payload) != "world" {
				t.Fatalf("pmessage=%q %q %v", ch, payload, rerr)
			}
		})
	}
}

func TestAnonymousMemSubscriberRejected(t *testing.T) {
	// Embedded-only: it opens the in-process engine directly.
	if openEmbeddedStore == nil {
		t.Skip("no embedded engine in this build")
	}
	_, err := SubscriberConnect("mem://")
	if !IsKind(err, KindUnsupported) {
		t.Fatalf("anonymous mem:// Subscriber should be Unsupported, got %v", err)
	}
}

func TestRemoteHello3PushFrames(t *testing.T) {
	s := spawnServer(t)
	sub, err := SubscriberConnect(s.url())
	if err != nil {
		t.Fatal(err)
	}
	defer sub.Close()
	// HELLO 3 must precede subscribe.
	if _, err := sub.Hello3(); err != nil {
		t.Fatalf("Hello3: %v", err)
	}
	if err := sub.Subscribe(b("rt")); err != nil {
		t.Fatal(err)
	}
	if ev, err := sub.Recv(); err != nil || ev.Kind != EventSubscribe {
		t.Fatalf("subscribe ack under RESP3: %+v %v", ev, err)
	}

	pub := mustConnect(t, s.url())
	time.Sleep(50 * time.Millisecond)
	_, _ = pub.Publish(bg, b("rt"), b("v3"))
	// recv handles RESP3 push (>N) transparently.
	ch, payload, rerr := sub.RecvMessage()
	if rerr != nil || string(ch) != "rt" || string(payload) != "v3" {
		t.Fatalf("RESP3 push delivery=%q %q %v", ch, payload, rerr)
	}
}

func TestEmbeddedHello3Unsupported(t *testing.T) {
	// Embedded-only: it opens the in-process engine directly.
	if openEmbeddedStore == nil {
		t.Skip("no embedded engine in this build")
	}
	sub, err := SubscriberConnect("mem://h3-bus")
	if err != nil {
		t.Fatal(err)
	}
	defer sub.Close()
	if _, err := sub.Hello3(); !IsKind(err, KindUnsupported) {
		t.Fatalf("embedded Hello3 should be Unsupported, got %v", err)
	}
}

func TestSubscriberReadTimeout(t *testing.T) {
	for _, bk := range pubsubBackends(t) {
		t.Run(bk.name, func(t *testing.T) {
			url, cleanup := bk.url(t)
			defer cleanup()
			sub, err := SubscriberConnectChannels(url, b("quiet"))
			if err != nil {
				t.Fatal(err)
			}
			defer sub.Close()
			_, _ = sub.Recv() // drain ack
			d := 60 * time.Millisecond
			if err := sub.SetReadTimeout(&d); err != nil {
				t.Fatal(err)
			}
			start := time.Now()
			_, rerr := sub.Recv()
			if rerr == nil {
				t.Fatal("expected a timeout error on a quiet channel")
			}
			if time.Since(start) > 2*time.Second {
				t.Fatal("read timeout did not bound the blocking recv")
			}
		})
	}
}
